src/cgw_ucentral_messages_queue_manager.rs

```diff
@@ -0,0 +1,360 @@
+        let container_lock = self.queue.read().await;
+
+        if self.get_device_messages_queue_len(device_mac).await == 0 {
+            return None;
+        }
+
+        let mut device_msg_queue = container_lock.get(device_mac).unwrap().write().await;
+        let reboot_msg = device_msg_queue
+            .get_item(MESSAGE_QUEUE_REBOOT_MSG_INDEX)
+            .unwrap()
+            .clone();
+        let configure_msg = device_msg_queue
+            .get_item(MESSAGE_QUEUE_CONFIGURE_MSG_INDEX)
+            .unwrap()
+            .clone();
+
+        if reboot_msg.parsed_message != default_msg {
+            ret_msg = reboot_msg;
+            device_msg_queue.remove_item(MESSAGE_QUEUE_REBOOT_MSG_INDEX);
+            device_msg_queue.insert_item(
+                MESSAGE_QUEUE_REBOOT_MSG_INDEX,
+                CGWUCentralMessagesQueueItem::default(),
+            );
+        } else if configure_msg.parsed_message != default_msg {
+            ret_msg = configure_msg;
+            device_msg_queue.remove_item(MESSAGE_QUEUE_CONFIGURE_MSG_INDEX);
+            device_msg_queue.insert_item(
+                MESSAGE_QUEUE_CONFIGURE_MSG_INDEX,
+                CGWUCentralMessagesQueueItem::default(),
+            );
+        } else {
+            ret_msg = device_msg_queue
+                .remove_item(MESSAGE_QUEUE_OTHER_MSG_INDEX)
+                .unwrap();
+        }
+
+        Some(ret_msg)
+    }
+
+    pub async fn get_device_queue_state(
+        &self,
+        device_mac: &String,
+    ) -> CGWUCentralMessagesQueueState {
+        let container_lock = self.queue.read().await;
+        let device_msg_queue = container_lock.get(device_mac).unwrap().read().await;
+
+        device_msg_queue.get_state()
+    }
+
+    pub async fn set_device_queue_state(
+        &self,
+        device_mac: &String,
+        state: CGWUCentralMessagesQueueState,
+    ) {
+        let container_lock = self.queue.read().await;
+        let mut device_msg_queue = container_lock.get(device_mac).unwrap().write().await;
+
+        device_msg_queue.set_state(state);
+    }
+
+    pub async fn insert_disconnected_device_timeout(&self, device_mac: String, seconds: u64) {
+        let mut disconnected_lock = self.disconnected_devices.write().await;
+        disconnected_lock.insert(device_mac, Duration::new(seconds, 0));
+    }
+
+    pub async fn remove_disconnected_device_timeout(&self, device_mac: &String) {
+        let mut disconnected_lock = self.disconnected_devices.write().await;
+        disconnected_lock.remove(device_mac);
+    }
+
+    pub async fn iterate_over_disconnected_devices(&self) {
+        let container_read_lock = self.disconnected_devices.read().await;
+        let mut update_internal_list: Vec<String> = Vec::<String>::new();
+        // 1. Iterate over disconnected devices and
+        // remove message queue if it is already empty
+        // or decrement timeout value
+        for (device_mac, _) in container_read_lock.iter() {
```

Comment:
Less explicit (hidden logic / syntax sugar) but a more idiomatic way to remove elements based on condition is to use [retain_mut](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain_mut) (or retain variant, based on the usage).